### PR TITLE
[Resources] Add memory in resources

### LINKED ISSUE #1673: Support auto selecting instance types via `memory` filter
User requested:

> I saw that Sky recently added support for specifying cpu resource constraints. Would it be possible to add memory resource constraints too? The use case is that I’m using dev instance to do some data analysis that requires a certain amount of memory to be available.

The user manually selected `r6i.metal` w/ 1TB memory, 128 vCPUs. The user did not see that `x2gd.metal` (Graviton) with 1TB mem, 64 vCPUs would be much cheaper. On this suggestion the user +1'd the auto selection by memory. (NOTE: we don't currently support it but will soon; but the point is there's complexity involved in selecting the best feasible instance type.) 

The user confirmed that the main motivation is not have to worry about learning which instance type to pick.

cc @WoosukKwon

### PR DESCRIPTION
**Pull request by Michaelvll**
Fixes #1673, this simulates the behavior of `cpus` in resources, thanks to the great effort from @WoosukKwon in #1622 

Added a memory filtering field for the resources:
1. `memory: 16` the memory should be 16GB
2. `memory: 16+` the memory should be more than 16GB
3. ~~`memory: 4x` the memory to core ratio should be larger than 4 times.~~ (Not expose this API as it might be confusing)

Logics for choosing the CPU instance to use:
1. Use the general-purpose CPU instance by default, when the `--memory` is not specified
2. Use the cheapest CPU instance that satisfies the requirement, when the `--memory` is specified.

1. `sky launch -c normal`
```
I 03-07 21:32:21 optimizer.py:712] Considered resources (1 node):
I 03-07 21:32:21 optimizer.py:760] ------------------------------------------------------------------------------------------------
I 03-07 21:32:21 optimizer.py:760]  CLOUD    INSTANCE           vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE   COST ($)   CHOSEN   
I 03-07 21:32:21 optimizer.py:760] ------------------------------------------------------------------------------------------------
I 03-07 21:32:21 optimizer.py:760]  AWS      m6i.2xlarge        8       32        -              us-east-2     0.38          ✔     
I 03-07 21:32:21 optimizer.py:760]  Azure    Standard_D8_v5     8       32        -              eastus        0.38                
I 03-07 21:32:21 optimizer.py:760]  GCP      n2-standard-8      8       32        -              us-central1   0.39                
I 03-07 21:32:21 optimizer.py:760]  Lambda   gpu_1x_a100_sxm4   30      200       A100:1         us-east-1     1.10                
I 03-07 21:32:21 optimizer.py:760] ------------------------------------------------------------------------------------------------
```
2. `sky launch -c cpus --cpus 32+`: general purpose
```
I 03-07 21:33:55 optimizer.py:760] -----------------------------------------------------------------------------------------------
I 03-07 21:33:55 optimizer.py:760]  CLOUD    INSTANCE          vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE   COST ($)   CHOSEN   
I 03-07 21:33:55 optimizer.py:760] -----------------------------------------------------------------------------------------------
I 03-07 21:33:55 optimizer.py:760]  AWS      m6i.8xlarge       32      128       -              us-east-2     1.54          ✔     
I 03-07 21:33:55 optimizer.py:760]  Azure    Standard_D32_v5   32      128       -              eastus        1.54                
I 03-07 21:33:55 optimizer.py:760]  GCP      n2-standard-32    32      128       -              us-central1   1.55                
I 03-07 21:33:55 optimizer.py:760]  Lambda   gpu_4x_a6000      56      400       A6000:4        us-east-1     3.20                
I 03-07 21:33:55 optimizer.py:760] -----------------------------------------------------------------------------------------------
```
3. `sky launch -c compute-optimized --cpus 32+ --memory 32+`
```
I 03-07 21:43:40 optimizer.py:760] ------------------------------------------------------------------------------------------------
I 03-07 21:43:40 optimizer.py:760]  CLOUD    INSTANCE           vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE   COST ($)   CHOSEN   
I 03-07 21:43:40 optimizer.py:760] ------------------------------------------------------------------------------------------------
I 03-07 21:43:40 optimizer.py:760]  GCP      n2-highcpu-32      32      32        -              us-central1   1.15          ✔     
I 03-07 21:43:40 optimizer.py:760]  Azure    Standard_F32s_v2   32      64        -              eastus        1.35                
I 03-07 21:43:40 optimizer.py:760]  AWS      c6i.8xlarge        32      64        -              us-east-2     1.36                
I 03-07 21:43:40 optimizer.py:760]  Lambda   gpu_4x_a6000       56      400       A6000:4        us-east-1     3.20                
I 03-07 21:43:40 optimizer.py:760] ------------------------------------------------------------------------------------------------
I 03-07 21:43:40 optimizer.py:760] 
```
4. `sky launch -c memory-optimized --memory 32+`
```
I 03-07 21:44:13 optimizer.py:760] ------------------------------------------------------------------------------------------------
I 03-07 21:44:13 optimizer.py:760]  CLOUD    INSTANCE           vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE   COST ($)   CHOSEN   
I 03-07 21:44:13 optimizer.py:760] ------------------------------------------------------------------------------------------------
I 03-07 21:44:13 optimizer.py:760]  AWS      r6i.xlarge         4       32        -              us-east-2     0.25          ✔     
I 03-07 21:44:13 optimizer.py:760]  Azure    Standard_E4_v5     4       32        -              eastus        0.25                
I 03-07 21:44:13 optimizer.py:760]  GCP      n2-highmem-4       4       32        -              us-central1   0.26                
I 03-07 21:44:13 optimizer.py:760]  Lambda   gpu_1x_a100_sxm4   30      200       A100:1         us-east-1     1.10                
I 03-07 21:44:13 optimizer.py:760] ------------------------------------------------------------------------------------------------
```

Tested (run the relevant ones):

- [x] Any manual or new tests for this PR (please specify below)
  - [x] `pytest tests/test_optimizer_dryruns.py` with added memory related unit tests.
- [x] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: tests/test_optimizer_random_dag.py [hunk 1]
```diff
@@ -116,7 +116,7 @@ def compare_optimization_results(dag: sky.Dag, minimize_cost: bool):
                                                       dag.tasks, optimizer_plan)
 
     min_objective = find_min_objective(copy_dag, minimize_cost)
-    assert objective == min_objective
+    assert abs(objective - min_objective) < 1e-6
```
Comment (WoosukKwon) [inline tests/test_optimizer_random_dag.py:119]: Nice catch. I think it'd be convenient to user a larger tolerance here, e.g., 1e-3.

File: sky/clouds/gcp.py [hunk 2]
```diff
@@ -402,6 +408,20 @@ def get_feasible_launchable_resources(self, resources):
                     cpus = float(resources.cpus)
                     if cpus != num_cpus_in_tpu_vm:
                         return ([], fuzzy_candidate_list)
+            memory_in_tpu_vm = 624
+            if resources.memory is not None:
+                if resources.memory.endswith('+'):
+                    memory = float(resources.memory[:-1])
+                    if memory > memory_in_tpu_vm:
+                        return ([], fuzzy_candidate_list)
+                elif resources.memory.endswith(('x', 'X')):
```
Comment (WoosukKwon) [inline sky/clouds/gcp.py:417]: As it is only used internally, I think we can only allow 'x' and disallow 'X'. WDYT?

File: sky/resources.py [hunk 3]
```diff
@@ -447,29 +498,54 @@ def _try_validate_instance_type(self) -> None:
...
+                                f'memory, but {self.memory} is requested.')
+                elif self.memory.endswith(('x', 'X')):
+                    requested_mem = float(self.memory[:-1] * cpus)
+                    if mem < requested_mem:
+                        with ux_utils.print_exception_no_traceback():
+                            raise ValueError(
+                                f'{self.instance_type} does not have the '
+                                f'requested memory. {self.instance_type} has '
+                                f'{cpus} CPUsand {mem} GB memory, but '
+                                f'{self.memory} is requested.')
```
Comment (WoosukKwon) [inline sky/resources.py:542]: Do we need this?

File: sky/clouds/service_catalog/gcp_catalog.py [hunk 4]
```diff
@@ -195,61 +209,42 @@ def get_instance_type_for_accelerator(
...
+        df = _df[_df['InstanceType'].notna()]
+        instance_type = _A100_INSTANCE_TYPE_DICTS[acc_name][acc_count]
+        df = df[df['InstanceType'] == instance_type]
+
+        # Check the cpus and memory specified by the user.
+        instance_type = common.get_instance_type_for_cpus_mem_impl(
+            df, cpus, memory)
+        if instance_type is None:
+            return None, []
+        return [instance_type], []
```
Comment (WoosukKwon) [inline sky/clouds/service_catalog/gcp_catalog.py:229]: Nice!

**Review by WoosukKwon — APPROVED** [review]
Great job! Thanks for the PR. Left some minor comments.

File: sky/clouds/service_catalog/aws_catalog.py [hunk 5]
```diff
@@ -27,9 +27,12 @@
 
 # This is the latest general-purpose instance family as of Jan 2023.
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/clouds/service_catalog/aws_catalog.py]: I changed the info for CPU types and the dates for each family. Thanks for pointing it out!

File: sky/clouds/service_catalog/aws_catalog.py [hunk 6]
```diff
@@ -27,9 +27,12 @@
...
 # CPU: Intel Ice Lake 8375C.
-# Memory: 4 GiB RAM per 1 vCPU.
-_DEFAULT_INSTANCE_FAMILY = 'm6i'
+# Memory: m6i - 4 GiB RAM per 1 vCPU;
+#         r6i - 8 GiB RAM per 1 vCPU;
+#         c6i - 2 GiB RAM per 1 vCPU.
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/clouds/service_catalog/aws_catalog.py]: Thanks! I made the comment per family to make it clearer.

**Pull request merged**